wandb/sdk/launch/utils.py

```diff
@@ -50,38 +51,80 @@ def _is_wandb_local_uri(uri: str):
     return _WANDB_LOCAL_DEV_URI_REGEX.match(uri)
 
 
-def fetch_and_validate_project(
+def set_project_entity_defaults(uri, project, entity, api):
+    # set the target project and entity if not provided
+    if not _is_wandb_uri(uri):
+        wandb.termlog("Non-wandb path detected")
+    _, uri_project, run_id = parse_wandb_uri(uri)
+    if project is None:
+        project = api.settings("project") or uri_project or UNCATEGORIZED_PROJECT
+    if entity is None:
+        entity = api.default_entity
+    return project, entity, run_id
+
+
+def construct_run_spec(
     uri,
-    target_entity,
-    target_project,
     experiment_name,
-    api,
-    version,
+    wandb_project,
+    wandb_entity,
+    docker_image,
     entry_point,
+    version,
     parameters,
-    user_id,
-    docker_image,
-    run_config,
+    launch_config,
 ):
-    parameters = parameters or {}
-    experiment_name = experiment_name
-    project = _project_spec.Project(
+    # override base config (if supplied) with supplied args
+    run_spec = launch_config if launch_config is not None else {}
+    run_spec["uri"] = uri
+    run_spec["entity"] = wandb_entity
+    run_spec["project"] = wandb_project
+    run_spec["name"] = experiment_name
```

Comment:
won't this automatically overwrite the values in the run_spec even if the args aren't given?

Reply:
oh yeah you're right idk why i did that